File: Data_Structures/WithCsharp/LinkedList/Singly/SinglyList.cs
public class SinglyListNode<T>
{
    public SinglyListNode(T data)
    {
        Data = data;
        Next = null;
    }

    public SinglyListNode() {}

    public required T Data { get; set; }
    public required SinglyListNode<T>? Next { get; set; }


    public SinglyListNode<T> StartList(T data) =>
        new SinglyListNode<T>()
        {
            Data = data,
            Next = null,
        };
}

public class SinglyList<T>
{
    private SinglyListNode<T> _singlyList;
    public SinglyList()
    {
        _singlyList = new SinglyListNode<T>() { Data = default, Next = null};
    }

    public SinglyListNode<T> StartList(T data) =>
        _singlyList.StartList(data);

    public SinglyListNode<T> AddLast(T data)
    {
        SinglyListNode<T> newElement = new SinglyListNode<T>() { Data = data, Next = null };
        if (_singlyList == null)
            return _singlyList = newElement;

        while (_singlyList.Next != null)
        {
            _singlyList = _singlyList.Next;
        }

        _singlyList.Next = newElement;
        return _singlyList;
    }

    public T GetIndex(int index)
    {
        if (_singlyList == null)
            return default;

        var tempElement = _singlyList;

        if (index < 0)
            throw new ArgumentOutOfRangeException(nameof(index));

        for(int i = 0; _singlyList != null && i < index; i++ )
        {
            tempElement = tempElement.Next;
        }

        if (tempElement == null)
            throw new ArgumentOutOfRangeException(nameof(index));

        return tempElement.Data;
    }

    public int GetLenght()
    {
        if (_singlyList == null)
            return 0;

        var tempElement = _singlyList;

        int lenght = 1;
        while(tempElement != null)
        {
            tempElement = tempElement.Next;
            lenght++;
        }

        return lenght;
    }

    public List<T> GetListData()
    {
        if (_singlyList == null)
            return new List<T>();

        var tempElement = _singlyList;
        List<T> list = new();

        while (tempElement != null)
        {
            list.Add(tempElement.Data);
            tempElement = tempElement.Next;
        }

        return list;
    }

    public bool DeleteElement(T delete)
    {
        if (_singlyList == null)
            return false;

        var tempElement = _singlyList;

        while (tempElement != null)
        {
            if (tempElement.Data != null && tempElement.Data.Equals(delete))
            {
                tempElement = tempElement.Next;
                _singlyList = tempElement;
            }
        }
    }
}

File: Data_Structures/Rust/data_structures/src/linked_list/singly_list.rs
#[derive(Debug)]
struct SinglyListNode<T> {
    pub data: T,
    pub next: Option<Box<SinglyListNode<T>>>,
}

impl<T> SinglyListNode<T> {
    fn new(data: T) -> SinglyListNode<T> {
        Self {
            data: data,
            next: None,
        }
    }
}

#[derive(Debug)]
pub struct Head<T> {
    head: Option<Box<SinglyListNode<T>>>,
}

impl<T: PartialEq> Head<T> {
    pub fn new() -> Self {
        Self { head: None }
    }

    pub fn add_last(&mut self, data: T) {
        let new_element = Some(Box::new(SinglyListNode::new(data)));

        match self.head.as_mut() {
            None => {
                self.head = new_element;
            }
            Some(mut temp_element) => {
                while let Some(ref mut next_node) = temp_element.next {
                    temp_element = next_node;
                }
                temp_element.next = new_element;
            }
        }
    }

    pub fn get_index(&mut self, index: i32) -> Option<&T> {
        if index < 0 {
            return None;
        }

        match self.head.as_mut() {
            None => None,
            Some(mut temp_element) => {
                let mut i: i32 = 0;
                while i < index {
                    temp_element = temp_element.next.as_mut().unwrap();
                    i += 1;
                }
                Some(&temp_element.data)
            }
        }
    }

    pub fn get_lenght(&mut self) -> i32 {
        match self.head.take() {
            None => 0,
            Some(mut element) => {
                let mut lenght = 1;
                while let Some(next) = element.next {
                    element = next;
                    lenght += 1;
                }
                lenght
            }
        }
    }

    pub fn remove(&mut self, v: T) -> bool {
        let mut current = &mut self.head;
        loop {
            match current {
                None => return false,
                Some(node) if node.data == v => {
                    *current = node.next.take();
                    return true;
                },
                Some(node) => {
                    current = &mut node.next;
                }
            }
        }
    }
}

#[cfg(test)]
mod test_singly_list {
    use super::*;

    #[test]
    fn basics_i32() {
        let mut list: Head<i32> = Head::new();
        list.add_last(5);
        list.add_last(10);
        list.add_last(11);
        assert_eq!(list.get_index(0), Some(&5));
        assert_eq!(list.get_index(1), Some(&10));
        assert_eq!(list.get_index(2), Some(&11));
    }

    #[test]
    fn basics_string() {
        let mut list: Head<String> = Head::new();

        list.add_last("ABC".to_string());
        list.add_last("123".to_string());
        list.add_last("Teste".to_string());

        assert_eq!(list.get_index(0), Some(&"ABC".to_string()));
        assert_eq!(list.get_index(1), Some(&"123".to_string()));
        assert_eq!(list.get_index(2), Some(&"Teste".to_string()));
    }

    #[test]
    fn get_lenght() {
        let mut list = Head::new();
        list.add_last(5);
        list.add_last(6);
        list.add_last(7);
        list.add_last(8);
        assert_eq!(list.get_lenght(), 2);
    }

    #[test]
    fn remove_data() {
        let mut list = Head::new();
        list.add_last(5);
        list.add_last(6);
        list.add_last(7);
        list.add_last(8);
        assert_eq!(list.remove(7), true);
        assert_eq!(list.remove(10), false);
        assert_eq!(list.get_lenght(), 3);
    }
}
